src/controllers/krate/delete.rs

```diff
@@ -0,0 +1,420 @@
+use crate::app::AppState;
+use crate::auth::AuthCheck;
+use crate::models::{Crate, Rights};
+use crate::schema::{crate_downloads, crates, dependencies};
+use crate::util::errors::{crate_not_found, custom, AppResult, BoxedAppError};
+use crate::worker::jobs;
+use axum::extract::Path;
+use bigdecimal::ToPrimitive;
+use chrono::Utc;
+use crates_io_worker::BackgroundJob;
+use diesel::prelude::*;
+use diesel_async::scoped_futures::ScopedFutureExt;
+use diesel_async::{AsyncConnection, RunQueryDsl};
+use http::request::Parts;
+use http::StatusCode;
+
+const DOWNLOADS_PER_MONTH_LIMIT: u64 = 100;
+
+pub async fn delete(Path(name): Path<String>, parts: Parts, app: AppState) -> AppResult<()> {
+    let mut conn = app.db_write().await?;
+
+    // Check that the user is authenticated
+    let auth = AuthCheck::only_cookie().check(&parts, &mut conn).await?;
+
+    // Check that the crate exists
+    let krate: Crate = Crate::by_name(&name)
+        .first(&mut conn)
+        .await
+        .optional()?
+        .ok_or_else(|| crate_not_found(&name))?;
+
+    // Check that the user is an owner of the crate (team owners are not allowed to delete crates)
+    let user = auth.user();
+    let owners = krate.async_owners(&mut conn).await?;
+    match user.rights(&app, &owners).await? {
+        Rights::Full => {}
+        Rights::Publish => {
+            let msg = "team members don't have permission to delete crates";
+            return Err(custom(StatusCode::FORBIDDEN, msg));
+        }
+        Rights::None => {
+            let msg = "only owners have permission to delete crates";
+            return Err(custom(StatusCode::FORBIDDEN, msg));
+        }
+    }
+
+    // Check that the requirements for deleting the crate are met
+    //
+    // - The crate has been published for less than 72 hours,
+    // - or if all the following conditions are met:
+    //     - The crate has a single owner,
+    //     - The crate has been downloaded less than 100 times for each month it has been published.
```

Comment:
While the value is calculated as an average rather than the exact value per month, I think this is the best approach.

Reply:
yeah, at first I tried to implement it properly, but datetime-math is hard and I don't think the tradeoff for the additional complexity is worth it in this case :D

Reply:
And especially, we only keep 90 days of download count data in the database, with older data being archived. This means that if we want to implement this with an exact download count, we'll either need to store a maximum monthly download count or face a more complex and less worthwhile solution.